### DISCUSSION TITLE
In deepin-devicemanager, I can build it, but the deb seems not work well.

### DISCUSSION THREAD

**Discussion by utt54321** [linuxdeepin/developer-center#3297] [Q&A | 问答板块]
I can build it, but the deb seems not work well. There will be segfault in /var/log/message
I see the Readme.md mentioned that should build it with polkit, I think it is why I can not use the self-builded deepin-devicemanager,  where can I find
authenticateProxy/policy/com.deepin.pkexec.deepin-devicemanager-authenticateProxy.policy
and
build-all-unknown-Debug/authenticateProxy/deepin-devicemanager-authenticateProx

**Comment by utt54321** [comment]
关于deepin-devicemanager，我发现在github上有相关的库https://github.com/linuxdeepin/deepin-devicemanager
但是其中的changelog只有1.0-1，我使用这个git库clone之后编译，但是会报泛型相关的代码错误
此外，我发现另一个库https://github.com/UbuntuDDE/deepin-devicemanager，然后我使用这个git仓库编译（跳过了测试，在debian11中尝试），可以编译成功，但是会出现segfault导致dbus服务无法正常工作，我想可能原因是在Readme.md中提到的需要polkit的支持，但是对于其提到的
authenticateProxy/policy/com.deepin.pkexec.deepin-devicemanager-authenticateProxy.policy
和
build-all-unknown-Debug/authenticateProxy/deepin-devicemanager-authenticateProx
我完全没有头绪。**是否有大佬知道这两个文件在哪里获取**。

后来我又到163的镜像源中下载deepin-devicemanager的源码包，使用dpkg-source -x解开后，里面的control额外增加了两个依赖，分别是deepin-desktop-base，deepin-keyring
当安装这两个依赖时，编译出来的deepin-devicemanager可以正常显示信息
当使用dpkg --purge删除这两个依赖时，在control中去除两个依赖，编译出来的deepin-devicemanager也会出现上述类似的segfault报错，从而dbus服务无法正常工作。
我非常好奇这是为什么，我想要在不依赖deepin-desktop-base，deepin-keyring的情况下，编译出可以正常使用的deepin-devicemanager

**Comment by BLumia** [comment]
一些背景信息：

deepin 新增的一些软件很多都是完全在 deepin/UOS 发行版中进行研发与测试的，可能对应的研发组在研发过程完全没考虑可移植性，所以如果遇到打包移植的问题的话，我们会建议研发组那边修改。当然，也非常欢迎提交 Pull Request :)

p.s. 顺带一问，你是在试图移植 deepin 的软件包到 debian 或者 ubuntu 之类的发行版吗？

关于你的问题，概要的回答：

> 其中的changelog只有1.0-1

因为之前是根据 git tag 来在打包过程中自动更新 debian/changelog 的，所以很多项目没直接去更新那个 changelog 文件，你可以考虑根据获取到的最新（或者目标）git tag 来使用 dch 更新 changelog 来帮助打包。

> 在Readme.md中提到的需要polkit的支持

我没特别留意这个项目，但据目前来看，有的项目是依赖 deepin 发行版内魔改过的 polkit 的。这种做法其实不合适，如果是因为这个的话，后续会督促项目修改做法。

> ...这两个文件在哪里获取。

我在一个 deepin 环境下并没有找到同名文件（已安装 deepin-devicemanager），可能 UbuntuDDE 仓库的包比较老吧，可以确认一下。另外，`build-all-unknown-Debug/` 是本地构建才会有的目录...

> deepin-desktop-base，deepin-keyring

keyring 那个包应该没什么值得一提的东西，desktop-base 里我记得有一些发行版相关的配置，没准 devicemanager 有依赖一些检查（当然，如果存在的话，实际上也不应该做这种检查）。

> segfault报错

可以考虑 coredumpctl 看下崩溃时的调用栈。不过，另一方面，如果程序本身跑起来都会有问题的话，我更建议先正常以调试的形式编译然后尝试运行，看看崩溃到哪里了，把这些问题确认一下后再着手处理打包。

**Reply by utt54321** [reply]
我是在debian11上尝试的，我总觉得开源软件应当在所有linux上可以进行使用，毕竟平台都是linux，而不是说跨平台到Android或Mac。
deepin-devicemanager里的内容是QT5 支持的，我现在想要在debian11上编译并且运行，但是遇到情况是：

> 到163的镜像源中下载deepin-devicemanager的源码包，使用dpkg-source -x解开后，里面的control额外增加了两个依赖，分别是deepin-desktop-base，deepin-keyring
> 当安装这两个依赖时，编译出来的deepin-devicemanager可以正常显示信息
> 当使用dpkg --purge删除这两个依赖时，在control中去除两个依赖，编译出来的deepin-devicemanager也会出现上述类似的segfault报错，从而dbus服务无法正常工作

这使我感到非常困惑不解，我自己去查看了deepin-desktop-base，deepin-keyring并没有发现非常特别的东西，真的是非常奇怪。
大佬，你能帮忙看一下吗？
此外，我发现使用dpkg --r删除deepin-desktop-base，deepin-keyring，也就是说保留配置信息，此时编译出来的deepin-devicemanager也能够正常使用，这让我更加凌乱了。
还有一点，因为我不会coredumpctl，我自己去打印日志看是在那里发生的段错误，我发现居然是在deepin-devicemanager-server的main函数里的a.exe()中
![image](https://user-images.githubusercontent.com/20319237/178093278-4d45cf57-50af-426a-bca9-2b853bb90eb7.png)
经过打印后的加了我的日志的段错误在syslog中如下图
![image](https://user-images.githubusercontent.com/20319237/178093377-4301a8c9-7c46-44b6-b181-2e2bd094066f.png)
可以看到我的“b2”的打印正好在segfault之前，这使得我更加更加凌乱了。求求大佬帮忙一起看看，球球了。

**Reply by BLumia** [reply]
我手头目前没有 debian 环境（现在是周末也不太想去折腾= =||），可能晚些时候会考虑在比较近似的环境里尝试一下。

> 我总觉得开源软件应当在所有linux上可以进行使用，毕竟平台都是linux，而不是说跨平台到Android或Mac。

写代码的时候也要注意可移植性才能做到跨平台（哪怕只是跨发行版）可用的。比如如果写了硬依赖 `apt` 包管理的代码就显然没法在非 debian 系发行版正常使用，写死了安装位置的代码在文件组织方式稍有不同的发行版里就无法正常工作之类。尽管 devicemanager 的问题应该不是这种程度的，但应该也属于可移植性的问题。

> 使用dpkg --r删除deepin-desktop-base，deepin-keyring，...，此时编译出来的deepin-devicemanager也能够正常使用，这让我更加凌乱了。

上面提到过了，这两个包提供的基本都是系统配置相关的内容，依赖是 debian/control 文件定义的，并不一定是写了就一定是没有就用不了的程度。

> 不会coredumpctl，我自己去打印日志看是在那里发生的段错误，我发现居然是在deepin-devicemanager-server的main函数里的a.exe()中

你很可能没有安装/提供对应软件包的调试信息（dbgsym 调试符号），所以能追到的就只有那里了。coredumpctl 其实蛮简单的，可以看下 https://bbs.deepin.org/post/235642 或者这个包的 man page。另外也可以尝试从 IDE 里进行编译和调试。

**Reply by utt54321** [reply]
（现在是周末也不太想去折腾= =||）

> 大佬，救人一命胜造七级浮屠，球球了，跪谢。QAQ

**Reply by utt54321** [reply]
大佬，我参照你说的链接用了coredumpctl，可以使报了如下的错：
![image](https://user-images.githubusercontent.com/20319237/178094215-30eea6ae-39e6-4f0e-9a98-bbb30e763415.png)

**Reply by BLumia** [reply]
:joy: 没 gdb 就装 gdb 呗... (p.s. 不用大佬大佬的叫的...)

**Reply by utt54321** [reply]
可是我tab 一下gdb可以看到有gdb
![image](https://user-images.githubusercontent.com/20319237/178094418-4b7a81e5-8240-40c2-9523-8306d46f15bb.png)

**Reply by BLumia** [reply]
只 `coredumpctl debug 1698649` 也不行对吗？

**Reply by utt54321** [reply]
![image](https://user-images.githubusercontent.com/20319237/178094496-0fae90b8-d4a7-4543-8892-9ec023cd1dcd.png)
有了，可是最后还是有报错，搞不懂

**Reply by utt54321** [reply]
这个堆栈似乎在主进程在main函数里调用qt5的exec的时候出错了，和我打印log死掉的地方一样，但是这QT5的堆栈我看不懂，我从来没学过qt5

**Reply by BLumia** [reply]
试了一下，在一台 ubuntu 22.04 环境下可以复现。应该是别的线程的问题，不是主线程，比预想的情况稍微复杂一些。实在不想周末调 bug，回头再看吧 :joy:

**Reply by utt54321** [reply]
比预想的情况稍微复杂一些

> QAQ

**Reply by utt54321** [reply]
大哥，我今天搜了一下午的百度，没有找到好的解释或类似的问题可以参考，有没有比较懂这个软件的贡献者可以参与看看。QAQ
我只知道poll是轮询时会用到，但是我几乎没有接触过这类东西，我很难理解其他线程poll在那里就停止的原因。

**Reply by BLumia** [reply]
考虑到目前看上去问题可能比较复杂了，并且可复现，可以考虑在 issue 面板开个 issue 来反馈这个 bug 了。

不过这类问题的处理优先级不见得高就是了，不见得会优先处理（其实我更倾向于先解决能让 debian/ubuntu/archlinux 跑 DDE 会遇到的问题，其次是 deepin 周边的非 DDE 相关的应用程序），但无论如何建议先开 bug，至少会有跟进状态

**Reply by utt54321** [reply]
我很想要解决这个问题，迫切需要，如果是写这个软件的贡献者参与会不会很快就找到问题所在，因为我想以前的版本不一定会有deepin-desktop-base，deepin-keyring这两个依赖，我想一定是在某个版本加入的，那应该可以看到当时的提交是否有配合这两个依赖的修改。

**Reply by BLumia** [reply]
另外我记得之前也有一些其它的与 devicemanager 移植相关的讨论，比如 https://github.com/linuxdeepin/developer-center/issues/3052 ，其它的可以在 https://github.com/linuxdeepin/developer-center/issues/ 查找（开 bug 也是在这里开）

**Reply by BLumia** [reply]
> 写这个软件的贡献者参与

目前 deepin 各个应用项目的研发人员都是 UnionTech 的员工（p.s. 我也是），项目研发人员周末是休息的，并且项目研发人员（至少目前）并不会优先处理非 deepin/UOS 上的问题，所以迫切需要的话恐怕只能靠自己了。

不过 devicemanager 并不是什么特殊的刚需性质的软件，为什么会迫切需要呢，基本需求的话其实可以考虑 lshw 之类的直接处理，或者你也可以尝试构建一下 devicemanager 的历史版本，兴许之前的版本是能用的。

> 以前的版本不一定会有deepin-desktop-base，deepin-keyring这两个依赖

我印象里是一直都依赖的。

> 我想一定是在某个版本加入的，那应该可以看到当时的提交是否有配合这两个依赖的修改。

可以试试看 `git blame` :)

**Reply by utt54321** [reply]
请问一下，在哪里可以获得历史版本的 devicemanager 呀

**Reply by BLumia** [reply]
https://github.com/linuxdeepin/deepin-devicemanager/tags 可以看历史 git tag。另外 https://github.com/linuxdeepin/developer-center/issues/3052 应该就是直接对应到你目前遇到的情况的（devicemanager-server 启动后会段错误）。

**Reply by utt54321** [reply]
QAQ 3052似乎最后没有得到解决，诶，心都碎了

**Reply by BLumia** [reply]
优先级不高，相比更高优先级的问题还比较多。如果有精力的话可以继续追一下问题，我们会尽量提供相关问题的支持（定位和解决问题过程中遇到的问题建议移步 #3052 讨论），但并不保证会优先处理这个问题。

**Reply by utt54321** [reply]
没救了，我曾经在一个开源项目pinta上解决过一个问题，花了我接近一周的时间，最后发现这个问题已经7年了。。。。。。从提出到我解决整整7年的时间，看来只能靠自己了，可是qt好难啊

**Reply by BLumia** [reply]
我还是不太理解为什么会有在别的发行版迫切需要使用 devicemanager 的需求...不过总之欢迎参与进来 :)

**Reply by utt54321** [reply]
因为我的老板要用这个软件，觉得好看QAQ

**Reply by utt54321** [reply]
在cmake的lists中有如下图的设置
![image](https://user-images.githubusercontent.com/20319237/178693621-1d9fbd53-8f53-4770-b10b-543fdf61f96a.png)
请问图中的COMMUNITY参数是什么，可以在哪里查看它的详细信息与使用方法？

**Reply by BLumia** [reply]
就是一个很常规的 cmake option，实质是一个缓存变量，通过 cmake gui / `ccmake` 或者执行 cmake 时通过 `-D` 方式传递定义，详见 cmake 的文档。

你截图所示位置就是对应到这个项目的相关的逻辑了，如果设为 ON 就会包含 COMMUNITY 宏定义，然后你在代码里搜索那个定义就是了。

**Reply by utt54321** [reply]
我使用163的deepin的源，下载的deepin-devicepackage
在编包过程中，在server的CMakeLists中将上述COMMUNITY的开关打开（原来是会根据系统环境中是否有/etc/os-release而打开，现在忽略这个条件，直接一直打开），之后编出的包在去除依赖deepin-desktop-base和deepin-keyring的情况下就可以正常打开显示信息了，不会报coredump
但是整个代码里COMMUNITY这个宏好像只在如下的代码里出现
![image](https://user-images.githubusercontent.com/20319237/178888053-17bd7afe-1857-4eb9-a452-d939fb264635.png)
我无法理解图中代码和coredump的关系
